test/unit/output/ferc1_test.py

```diff
@@ -0,0 +1,119 @@
+"""Tests for the FERC Form 1 output functions.
+
+Calculation tree generation:
+- [x] Test base recusion case (leaf node w/ empty calculations)
+- [x] Test simple 1-level recursion w/ 1-2 child nodes.
+- [x] Test calculation with non-unique source table (should fail)
+- Test calculation that references an outside table (???)
+
+"""
+
+import json
+import logging
+from contextlib import nullcontext as does_not_raise
+
+import pandas as pd
+import pytest
+
+from pudl.output.ferc1 import Ferc1XbrlCalculationTree
+
+logger = logging.getLogger(__name__)
+
+EXPLODED_META_IDX = ["table_name", "xbrl_factoid"]
+TEST_CALC_1 = [
+    {"name": "reported_1", "weight": 1.0, "source_tables": ["table_1"]},
+    {"name": "reported_2", "weight": -1.0, "source_tables": ["table_1"]},
+]
+
+TEST_CALC_2 = [
+    {"name": "reported_1", "weight": 1.0, "source_tables": ["table_1", "table_2"]},
+    {"name": "reported_2", "weight": -1.0, "source_tables": ["table_1"]},
+]
+
+TEST_CALC_3 = [
+    {"name": "reported_1", "weight": 1.0, "source_tables": ["table_1"]},
+    {"name": "reported_3", "weight": 1.0, "source_tables": ["table_3"]},
+]
```

Comment:
@cmgosnell @e-belfer 

I'm not sure how we want to handle this case, where the calculation refers to a table that is not in the set of tables being "exploded." This means the calculation subcomponents can't be looked up in the `exploded_meta` dataframe, which currently results in a `KeyError` while trying to build the tree.

Maybe we should catch that `KeyError` and verify that it's failing because the table is missing (rather than e.g. a typo in the `xbrl_factoid`) and return an empty calculation? Will that result in just looking up the "reported" calculated value rather than attempting to calculate the value that's out-of-bounds?

Reply:
To clarify, in this case one but not all of the components comes from outside of the set of exploded tables? 

Re: your second question - if this process is upstream of the reconciliation process and it returns an empty calculation, I think it should just return the reported value. The metadata will still mark it as `calculated_value` though, so we should think about whether we want to flag this in any way. 

I think the question of having some components sourced from in-table and some from out-of-table is discussed in #2622. For now, it seems like this isn't behavior that we're expecting to happen, so I think catching the `KeyError`, verifying it's due to an out-of-bounds table and returning either a warning ("unexpected components, component will not be calculated") or an error would make sense as the behavior here. I'd actually lean towards an error, though perhaps this will make things overly brittle.

Reply:
I don't have a lot of clarify around how the metadata explosion/calculation tree treatment is going to interplay with the data explosion. but i do feel clear-ish about how this type of thing needs to be treated in the output.

If any component of a calculated factoid is outside of the explosion tables that we care about we remove all of the components of the calculation *that are within the explosion* so we can leave only the calculated value. And then that calculated factoid is now the leaf.

So in the context of a factoid's calculation tree... idk. Removing the calculation itself and redefining the factoid as a non-calculated value would effectively remove the information about how to remove it's remaining sub-components. UNLESS we use the calc trees to compile the set of leaves we actually care about?

Reply:
Based on our conversation this morning, I think what we decided to do here is pre-emptively any calculation that mentions a table outside of the set of tables being exploded in the `ExplodeMeta` step, replacing it with `"[]"` and changing `xbrl_row_type` to `reported_value` -- for the purposes of the exploded table analysis.

Then, the calculation tree which is generated from the exploded metadata will automatically consider those values leaves, and any other values within the exploded tables that were referenced in mixed calculations (those with values from more than one table) will automatically get pruned from the overall calculation tree, making it very straightforward to use the calculation tree to identify the facts which need to be retained in the exploded dataframe (i.e. only those facts which are associated with leaves of the tree).